bundles/catalogue/metadatasearch/MetadataStateHandler.js

```diff
@@ -0,0 +1,260 @@
+import { StateHandler, controllerMixin } from 'oskari-ui/util';
+import { renderMetadataSearchContainer } from './view/MetadataSearchContainer';
+import { MetadataOptionService } from './service/MetadataOptionService';
+import { MetadataSearchService } from './service/MetadataSearchService';
+import { METADATA_BUNDLE_LOCALIZATION_ID } from './instance';
+
+class MetadataStateHandler extends StateHandler {
+    constructor (instance) {
+        super();
+        this.instance = instance;
+        this.optionsService = new MetadataOptionService(this.instance.optionAjaxUrl);
+        this.searchService = new MetadataSearchService(this.instance.searchAjaxUrl);
+        this.mapLayerService = this.getSandbox().getService('Oskari.mapframework.service.MapLayerService');
+        this.setState({
+            query: '',
+            advancedSearchExpanded: false,
+            advancedSearchOptions: null,
+            advancedSearchValues: {
+            },
+            loading: false,
+            drawing: false,
+            searchResultsVisible: false,
+            searchResultsFilter: null,
+            coverageFeature: null,
+            selectedLayers: null
+        });
+        this.addStateListener(() => this.updateMetadataSearch());
+    }
+
+    getSandbox () {
+        return this.instance.getSandbox();
+    }
+
+    getVectorLayerId () {
+        return this.instance.getVectorLayerId();
+    }
+
+    renderMetadataSearch (element) {
+        this.contentController = renderMetadataSearchContainer(this.getState(), this.getController(), element);
+    }
+
+    updateMetadataSearch () {
+        if (!this.contentController) {
+            this.contentController = renderMetadataSearchContainer(this.getState(), this.getController(), document.createElement('div'));
+        }
+
+        this.contentController.update(this.getState(), this.getController());
+    }
+
+    updateQuery (query) {
+        this.updateState({ query });
+    }
+
+    doSearch () {
+        this.updateState({
+            loading: true,
+            searchResultsVisible: false,
+            searchResultsFilter: null
+        });
+        const { query, advancedSearchValues } = this.getState();
+        const formdata = {};
+        if (query) {
+            formdata.search = query;
+        }
+
+        Object.keys(advancedSearchValues).forEach(key => {
+            if (advancedSearchValues[key] instanceof Array) {
+                formdata[key] = advancedSearchValues[key].join(',');
+            } else {
+                // eliminate nulls
+                if (advancedSearchValues[key]) {
+                    formdata[key] = advancedSearchValues[key];
+                }
+            }
+        });
+
+        this.searchService.doSearch(formdata, (results) => this.updateSearchResults(results));
+    }
+
+    updateSearchResults (json) {
+        json?.results?.forEach(result => this.updateLayerInfo(result));
```

Comment:
Maybe map() instead to return a modified result so it's easier to follow the code.

Reply:
yeah, definitely easier to read that way